### PR TITLE
Added check and test for betas parameter for Adam Optimizer

### PR DESCRIPTION
**Pull request by lazypanda1**
This PR adds a check to prevent division by zero errors and give users friendlier error messages when using the Adam optimizer.

Currently, if one specifies the beta value of the Adam optimizer as `1.0` for the first parameter, pyro fails with the error message, `ZeroDivisionError: float division by zero`. According to the definition of [Adam](https://arxiv.org/abs/1412.6980), beta values should be in the range \[0, 1\).

### CODE REVIEW

**Comment by martinjankowiak** [comment]
hello. thanks for the contribution! since we just wrap the adam in pytorch i think a more appropriate place for this kind of check would be in pytorch?

**Comment by fritzo** [comment]
:+1: I agree the PyTorch folks would appreciate better error checking, then a larger community could benefit from this fix.

File: pyro/optim/__init__.py [hunk 1]
```diff
@@ -11,8 +11,13 @@ def Adam(optim_args):
     """
     A wrapper for torch.optim.Adam
     """
-    return PyroOptim(torch.optim.Adam, optim_args)
+    return PyroOptim(torch.optim.Adam, optim_args, _adam_checker)
 
+def _adam_checker(optim_args):
+    assert (optim_args['betas'][0] >= 0.0 and optim_args['betas'][0] < 1.0 \
```
[the remark this answers is not in the document]
Comment (lazypanda1) [reply pyro/optim/__init__.py]: Done

File: pyro/optim/optim.py [hunk 2]
```diff
@@ -14,7 +14,7 @@ class PyroOptim(object):
     :param optim_args: a dictionary of learning arguments for the optimizer or a callable that returns
         such dictionaries
     """
-    def __init__(self, optim_constructor, optim_args):
+    def __init__(self, optim_constructor, optim_args, checker=None):
```
[the remark this answers is not in the document]
Comment (lazypanda1) [reply pyro/optim/optim.py]: Done

**Comment by fritzo** [comment]
@lazypanda1 After discussing with the rest of the Pyro team, we feel fix really belongs upstream in PyTorch so more people can benefit. The PyTorch folks are very friendly about accepting PRs (we contribute regularly). I think you can simply put your `ValueError`s now in `Adam.__init__()`: https://github.com/pytorch/pytorch/blob/master/torch/optim/adam.py#L29 If you send a PR this week it will likely make it into PyTorch 0.4 release and be available in the next major release of Pyro.

**Pull request closed without merging**